### ISSUE TITLE
Server skips messages to client makes the json-patch-queue wait forever

### ISSUE THREAD

**Issue by malx122** [Starcounter/Starcounter#3981]
Hi!
I started this issue at https://github.com/PuppetJs/PuppetJs/issues/108 but starting to realize that this is probably a server bug and not a puppetjs issue.

When two "internal" puppetjs requests from the client is sent faster than the server has time to respond the first one is omitted. Not sure if this is by design to avoid choking but this means that the _ver$s in the patch jumps over a version number and the client waits for the missing one forever, never applying any new patch.

Here I have triggered many requests on a slow remote server working with other heavy background processes in the same time making it easy to get the error:
![patches](https://cloud.githubusercontent.com/assets/913374/20628447/460e0698-b326-11e6-8485-3b18b9d9c61d.png)

**Comment by miyconst** [comment]
@chrhol, @bigwad, this seems to be the exactly same issue I have discovered and reported with my stress test app.

**Comment by miyconst** [comment]
>... is probably a server bug and not a puppetjs issue.

@malx122, yes, this is a server side bug. Please close the PuppetJs issue as duplicate.

**Comment by dan31** [comment]
@malx122 and @miyconst thanks for your thorough investigation!
@chrhol let's take it high prio and try to have it in 2.3, okay?
CC @tomalec @alemoi

**Comment by malx122** [comment]
> @malx122, yes, this is a server side bug. Please close the PuppetJs issue as duplicate.
I actually think this is an issue at the puppetjs side as well. To just stop doing things forever without trying to do something about it or at least notify the user is an error in my world. It shouldn't have taken me this long to track it down.

**Comment by chrhol** [comment]
Something else is happening here, I cannot see any problems with versions of patches being mixed up as they were in the case of @miyconst stresstest. 

I reproduced the issue locally by adding a delay in the patchgeneration to slow things down a bit and clicking on the menu-choices which sends http-request. I cannot reproduce it when doing stuff that sends over websocket.

I've added logging in the beginning when patchgeneration is started (i.e. when the version is updated) and another log when the response is sent (at the end in the managed code). 

In the debug-aid I get the same as @malx122 reported, where clearly patches are missing:
![debugaid](https://cloud.githubusercontent.com/assets/1658979/20664818/2714180c-b55d-11e6-8f78-7f34bd2d9a88.JPG)

But the log looks like this (`START` is starting patchgeneration and `SEND` is after managed code have run and response sent):
```
START >> Time: 2016-11-28T11:18:16.9169609+01:00, Version: 1, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:16.9789859+01:00, Version: 1, Thread: 35
START >> Time: 2016-11-28T11:18:19.3106890+01:00, Version: 2, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:19.3787097+01:00, Version: 2, Thread: 35
START >> Time: 2016-11-28T11:18:32.4835063+01:00, Version: 3, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:32.5485256+01:00, Version: 3, Thread: 35
START >> Time: 2016-11-28T11:18:40.0184500+01:00, Version: 4, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:40.0734701+01:00, Version: 4, Thread: 35
START >> Time: 2016-11-28T11:18:40.8740552+01:00, Version: 5, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:40.9095720+01:00, Version: 5, Thread: 35
START >> Time: 2016-11-28T11:18:47.4955271+01:00, Version: 6, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:47.5755573+01:00, Version: 6, Thread: 35
START >> Time: 2016-11-28T11:18:48.1856202+01:00, Version: 7, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:48.3063531+01:00, Version: 7, Thread: 35
START >> Time: 2016-11-28T11:18:56.9833547+01:00, Version: 8, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:57.0543850+01:00, Version: 8, Thread: 35
START >> Time: 2016-11-28T11:18:57.2569874+01:00, Version: 9, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:57.3300579+01:00, Version: 9, Thread: 35
START >> Time: 2016-11-28T11:18:57.5510747+01:00, Version: 10, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:57.6362916+01:00, Version: 10, Thread: 35
START >> Time: 2016-11-28T11:18:58.0226183+01:00, Version: 11, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:58.0851430+01:00, Version: 11, Thread: 35
START >> Time: 2016-11-28T11:18:58.2873709+01:00, Version: 12, Thread: 34
SEND (http) >> Time: 2016-11-28T11:18:58.3659030+01:00, Version: 12, Thread: 34
START >> Time: 2016-11-28T11:18:58.5688338+01:00, Version: 13, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:58.6383611+01:00, Version: 13, Thread: 35
START >> Time: 2016-11-28T11:18:58.8410952+01:00, Version: 14, Thread: 34
SEND (http) >> Time: 2016-11-28T11:18:58.9221217+01:00, Version: 14, Thread: 34
START >> Time: 2016-11-28T11:18:59.1245086+01:00, Version: 15, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:59.1850327+01:00, Version: 15, Thread: 35
START >> Time: 2016-11-28T11:18:59.5813520+01:00, Version: 16, Thread: 35
SEND (http) >> Time: 2016-11-28T11:18:59.6513804+01:00, Version: 16, Thread: 35
```

Somewhere along the way, the responses for patches with version 11-15 is lost.

**Comment by chrhol** [comment]
So I guess this might be a problem:
![canceled](https://cloud.githubusercontent.com/assets/1658979/20666154/21de1fbc-b563-11e6-9da1-f7d144d82c33.JPG)

When clicking very fast and the server is slow to respond, some requests is canceled in the browser. However they are not canceled on the server (since we have no way of doing that today), which leads to that the server still sends the response and updates the version.

I'm a bit out of my depth here when it comes to how browsers work. Is this some built-in behaviour or is this something we do in puppet.js? And how do we handle these cases.

@miyconst, @warpech, @tomalec

**Comment by miyconst** [comment]
>I'm a bit out of my depth here when it comes to how browsers work. Is this some built-in behaviour or is this something we do in puppet.js? And how do we handle these cases.

No, this is not browser behavior. It should not cancel the requests on its own. It looks to be the PuppetJs issue. 

@joozek78, this is something you have to take a look at. Only heart beat requests can be cancelled.

**Comment by warpech** [comment]
> When clicking very fast and the server is slow to respond, some requests are canceled in the browser. 

This happens exactly in line: https://github.com/PuppetJs/PuppetJs/blob/gh-pages/src/puppet.js#L433

@joozek78 I remember we were talking about this. I suggested removal of this line. Do you remember what's the reason it stayed?

**Comment by warpech** [comment]
> @joozek78 I remember we were talking about this. I suggested removal of this line. Do you remember what's the reason it stayed?

This was the issue: https://github.com/PuppetJs/PuppetJs/issues/93

**Comment by joozek78** [comment]
The reason it stayed was that it actually had little to do with the issue.

Probably it would be okay to remove that line, because [that line](https://github.com/PuppetJs/PuppetJs/blob/gh-pages/src/puppet.js#L355) already protects against duplicate WS channels (original reason for closing XHR). Of course that needs verification (probably with some crazy clicking to put the puppet off-balance). Unfortunately, I don't have the capacity to do that now.

**Comment by warpech** [comment]
@malx122 could you pls check if committing the line https://github.com/PuppetJs/PuppetJs/blob/gh-pages/src/puppet.js#L433 fixes the problem for you?

**Comment by malx122** [comment]
Seems to work for me! (There is still a puppet/starcounter-include issue that I got while testing but I think its unrelated)

**Comment by chrhol** [comment]
Cool, then it looks like removal of closing the request on the client will solve the original problems. I see that the changes are merged, but PuppetJs is not updated in Starcounter, right? @warpech, @tomalec

**Comment by warpech** [comment]
Yes, PuppetJs will be updated jointly with a fix for https://github.com/Starcounter/Starcounter/issues/3798

**Comment by warpech** [comment]
Assuming this fix can be made only in v2.3/pnext, just as #3798? Please confirm.

**Comment by warpech** [comment]
A decision was made that fix for #3981 and #3798 does not have to be in a single PuppetJs release.

The fix for #3981 is done in PuppetJs 2.3.0, which should be available in tomorrow Starcounter nightly build (`develop` and `pnext`).

**Comment by warpech** [comment]
Merged to `develop` and `pnext`.
